File: 0141.Linked-List-Cycle/step2.cpp
// Definition for singly-linked list.
#include <unordered_set>

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    bool hasCycle(ListNode *head) {
        std::unordered_set<ListNode *> seen;
        for (ListNode *node = head; node != nullptr; node = node->next) {
            if (seen.count(node)) {
                return true;
            }
            seen.insert(node);
        }
        return false;
    }
};

File: 0019.Remove-Nth-Node-From-End-of-List/step2.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: ListNode | None, n: int) -> ListNode | None:
        dummy = ListNode(next=head)
        slow = dummy
        fast = dummy
        for _ in range(n):
            fast = fast.next

        if fast is None:
            raise ValueError(f"n is too large: {n}")

        while fast.next:
            slow = slow.next
            fast = fast.next

        removed_node = slow.next
        slow.next = removed_node.next
        removed_node.next = None

        return dummy.next
